### ISSUE TITLE
NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence

### ISSUE THREAD

**Issue by chjohnst** [openzfs/zfs#6851]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | CentOS
Distribution Version    | 7.3
Linux Kernel                 | 3.10.0-514.26.2.el7.x86_64
Architecture                 | x86_64
ZFS Version                  | 0.6.5.9
SPL Version                  | 0.6.5.9

### Describe the problem you're observing
Repeated errors from my NFS clients on ZFS exported filesystems.  I have several non-ZFS exported filesystems that do not exhibit this error.  This found by looking in tcpdump and capturing for the error in particular.  (nfs.status == 10026)

[1008951.143700] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff8817c8fbee20!
[1008976.459382] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff88069e299020!
[1008976.520790] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff88069e29aa20!
[1008976.632506] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff88069e29ba20!
[1008976.843640] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff880080706220!
[1008977.254953] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff880080704620!
[1008980.346921] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff8800b25ed020!
[1008980.407991] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff8800b25eda20!
[1008980.519267] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff88057788e820!
[1008980.731117] NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence ffff88057788d420!

### Describe how to reproduce the problem
Mult-user system likely doing a lot of reads/writes and shared access to files

**Comment by vitlav** [comment]
I think it only NFS issue.

**Comment by paulraines68** [comment]
I have the same issue.  Same server (CentOS 7.5) has NFS exported ext4 and zfs (0.7.9) volumes.  This error only happens on the zfs volume.

**Comment by paulraines68** [comment]
It is unclear if the error/warning just has performance implications or if data integrity could be an issue.

**Comment by vincentkersten** [comment]
Just ran into this one, related to file locking. 
```
 zfs get sharenfs raid/home
NAME       PROPERTY  VALUE                                                                    SOURCE
raid/home  sharenfs  rw=@10.0.0.1/24,async,no_subtree_check,root_squash,insecure,crossmnt  local
```

in this case connected with svn use - no_subtree_check is set.
This did not happen sharing from ext4 before. 

Anyone knows a way out of this ?

**Comment by dhojelsen** [comment]
I am experiencing the same issues with a zfs based nfs export. I have several clients and running fine, but occasionally one of the nfs clients goes bad.

I get `NFS: v4 server returned a bad sequence-id error on an unconfirmed sequence` in the logs
I am able to access files but flock fails with an IO error.

I am running version 0.7.8-1

@vincentkersten, I am interested to know if you found a solution for your issue.

Thanks David

**Comment by vincentkersten** [comment]
@dhojelsen it so happens today i used maint-window to do the following:

-  moved back from backports kernel (4.19) to stable kernel (4.9 debian) . I used a backports kernel because th eZFS was not available before on the stable kernel. It is possible that the NFS kernel module in the 4.19 is different from the module in 4.9, causing havoc.
 
-  stayed on the zfs version from backports because a feature (org.zfsonlinux:userobj_accounting (User/Group object accounting.)) was not available in the stable kernel module (0.6.5 vs 0.7.12 from the top of my head) 

- added 'domain=my.domain.com'  as well as 'Cache-Expiration=10'  in /etc/idmapd.conf  (This is geared to LDAP and network setup)

I did not change any kernel-module parameters for now. Some seemed to be applicable to this situation, but documentation on some of them is not easy to find. 

I don't think it shows here, but i stopped using the ZFS-NFS (on nov 12th 2018, previous comment, already) and use the classic nfs-kernel-server.


For now, keeping a close eye on "lslocks" on the NFS server.
